src/hamt_state.rs

```diff
@@ -0,0 +1,174 @@
+
+        Self::ensure_owner(n)?;
+
+        map.delete(&key)?;
+        Ok(map.flush()?)
+    }
+}
+
+impl LoadableState for HamtState {
+    fn new(nodes: Vec<NodeInfo>) -> Result<Self, Error> {
+        let total_checkers = nodes.len();
+        let mut checker_map = make_empty_map::<_, NodeInfo>(&Blockstore);
+        for n in nodes {
+            let str = &n.creator().to_string()[..];
+            checker_map.set(BytesKey::from(str), n)?;
+        }
+        Ok(HamtState {
+            members: make_empty_map::<_, NodeInfo>(&Blockstore).flush()?,
+            checkers: checker_map.flush()?,
+            offline_checkers: make_empty_map::<_, Votes>(&Blockstore).flush()?,
+            total_checkers
+        })
+    }
+
+    fn upsert_node(&mut self, node: NodeInfo) -> Result<(), Error> {
+        self.members = Self::upsert(&Blockstore{}, &self.members, node)?;
+        Ok(())
+    }
+
+    fn remove_node(&mut self, id: &ActorID) -> Result<(), Error> {
+        self.members = Self::remove(&Blockstore{}, &self.members, id)?;
+        Ok(())
+    }
+
+    fn is_checker(&self, checker: &ActorID) -> Result<bool, Error> {
+        let map = get_map_from_cid::<_, NodeInfo>(&self.checkers, &Blockstore{})?;
+        let key = BytesKey::from(&checker.to_string()[..]);
+        Ok(map.contains_key(&key)?)
+    }
+
+    fn upsert_checker(&mut self, node: NodeInfo) -> Result<(), Error> {
+        self.checkers = Self::upsert(&Blockstore{}, &self.checkers, node)?;
+        Ok(())
+    }
+
+    fn remove_checker(&mut self, id: &ActorID) -> Result<(), Error> {
+        self.checkers = Self::remove(&Blockstore{}, &self.checkers, id)?;
+        Ok(())
+    }
+
+    fn remove_checker_unchecked(&mut self, checker: &ActorID) -> Result<(), Error> {
+        let mut map = get_map_from_cid::<_, NodeInfo>(&self.checkers, &Blockstore{})?;
+        map.delete(&BytesKey::from(&checker.to_string()[..]))?;
+        self.checkers = map.flush()?;
+        Ok(())
+    }
+
+    fn has_voted(&self, reported: &ActorID, voter: &ActorID) -> Result<bool, Error> {
+        let map = get_map_from_cid::<_, Votes>(&self.offline_checkers, &Blockstore{})?;
+        Ok(
+            map.get(&BytesKey::from(&reported.to_string()[..]))?
+                .map(|v| v.has_voted(voter))
+                .unwrap_or(false)
+        )
+    }
+
+    fn record_voted(&mut self, reported: &ActorID, voter: &ActorID) -> Result<usize, Error> {
+        let mut map = get_map_from_cid::<_, Votes>(&self.offline_checkers, &Blockstore{})?;
+        let reported_key = BytesKey::from(&reported.to_string()[..]);
+
+        match map.get(&reported_key)? {
+            None => {
+                let mut votes = Votes::new(fvm_sdk::network::curr_epoch());
+                votes.vote(voter);
+                map.set(reported_key, votes)?;
+                self.offline_checkers = map.flush()?;
+                Ok(1)
+            }
+            Some(votes) => {
+                let t = self.vote_threshold();
```

Comment:
Shouldn't we record the vote before checking if it is withing threshold and removing the checker instead of waiting for the next vote to remove it? 

(Maybe I haven't understood the code well, just to double-check)

Reply:
I updated this code already, there was a small issue when the voting duration is passed, new vote is not recorded in the new round. I have updated to recording already.